examples/chat.rs

```diff
@@ -0,0 +1,135 @@
+/// message is sent and printed in the other terminal. You can issue two special
+/// commands: `unsubscribe` and `subscribe` (to re-subscribe to the "chat"
+/// topic). Close with Ctrl-c.
+///
+/// You can of course open more terminal windows and add more participants.
+/// Dialing any of the other peers will propagate the new participant to all
+/// chat members and everyone will receive all messages.
+extern crate futures;
+extern crate libp2p;
+extern crate tokio;
+
+use futures::prelude::*;
+use libp2p::{Transport, core::upgrade, secio, mplex};
+use libp2p::tokio_codec::{FramedRead, LinesCodec};
+
+fn main() {
+    // Create a random PeerId
+    let local_key = secio::SecioKeyPair::ed25519_generated().unwrap();
+    let local_peer_id = local_key.to_peer_id();
+    println!("Local peer id: {:?}", local_peer_id);
+
+    // Set up a an encrypted DNS-enabled TCP Transport over the Mplex protocol
+    let transport = libp2p::CommonTransport::new()
+        .with_upgrade(secio::SecioConfig::new(local_key))
+        .and_then(move |out, endpoint| {
+            let peer_id = out.remote_key.into_peer_id();
+            let upgrade = upgrade::map(mplex::MplexConfig::new(), move |muxer| (peer_id, muxer));
+            upgrade::apply(out.stream, upgrade, endpoint.into())
+        });
+
+    // Create a Floodsub topic
+    let floodsub_topic = libp2p::floodsub::TopicBuilder::new("chat").build();
+
+    // Create a Swarm to manage peers and events
+    let mut swarm = {
+        let mut behaviour = libp2p::floodsub::FloodsubBehaviour::new(local_peer_id);
+        behaviour.subscribe(floodsub_topic.clone());
+        libp2p::Swarm::new(transport, behaviour, libp2p::core::topology::MemoryTopology::empty())
+    };
+
+    // Listen on all interfaces and whatever port the OS assigns
+    let addr = libp2p::Swarm::listen_on(&mut swarm, "/ip4/0.0.0.0/tcp/0".parse().unwrap()).unwrap();
+    println!("Listening on {:?}", addr);
+
+    // Reach out to another node
+    if let Some(to_dial) = std::env::args().nth(1) {
+        let dialing = to_dial.clone();
+        match to_dial.parse() {
+            Ok(to_dial) => {
+                match libp2p::Swarm::dial_addr(&mut swarm, to_dial) {
+                    Ok(_) => println!("Dialed {:?}", dialing),
+                    Err(e) => println!("Dial {:?} failed: {:?}", dialing, e)
+                }
+            },
+            Err(err) => println!("Failed to parse address to dial: {:?}", err),
+        }
+    }
+
+    // Read full lines from stdin
+    let stdin = tokio_stdin_stdout::stdin(0);
+    let mut framed_stdin = FramedRead::new(stdin, LinesCodec::new());
+
+    // Kick it off
+    tokio::run(futures::future::poll_fn(move || -> Result<_, ()> {
+        loop {
+            match framed_stdin.poll().expect("Error while polling stdin") {
+                Async::Ready(Some(line)) => {
+                    match line.as_ref() {
+                        "unsubscribe" => {
```

Comment:
I'm not sure we want to support the user writing `"unsubscribe"` and `"subscribe"`
It complicates things a bit, and isn't something you would do in a real-world application